docs/design-docs/08-package-variant.md

```diff
@@ -338,20 +337,91 @@ With PackageVariant, we instead create the following resources:
 - Annotations
 - Labels
 - Adoption and Deletion Policy
+- Status
+  - `Valid` Condition Type
+  - `DownstreamEnsured` Condition Type (`Ready`?)
+  - `DraftExists` Condition Type (?)
 
 #### Package Context Injection
 
+PackageVariant resource authors may specify key-value pairs in the
+`spec.packageContext.data` field of the resource. These key-value pairs will be
+automatically added to the `data` of the `kptfile.kpt.dev` ConfigMap, if it
+exists.
+
+Specifying the key `name` is invalid and must fail validation of the
+PackageVariant. This key is reserved for kpt or Porch to set to the package
+name.
+
+The `spec.packageContext.removeKeys` field can also be used to specify a list of
+keys that the package variant controller should remove from the `data` field of
+the `kptfile.kpt.dev` ConfigMap.
+
+When creating or updating a package, the package variant controller will ensure
+that:
+- The `kptfile.kpt.dev` ConfigMap exists (creating it if not)
+- All of the key-value pairs in `spec.packageContext.data` exist in the `data`
+  field of the ConfigMap.
+- None of the keys listed in `spec.packageContext.removeKeys` exist in the
+  ConfigMap.
+
+Note that if a user adds a key via PackageVariant, then changes the
+PackageVariant to no longer add that key, it will NOT be removed automatically,
+unless the user also lists the key in the `removeKeys` list. This avoids the
+need to track which keys were added by PackageVariant.
+
+Similarly, if a user manually adds a key in the downstream that is also listed
+in the `removeKeys` field, the package variant controller will remove that key
+the next time it needs to update the downstream package. There will be no
+attempt to coordinate "ownership" of these keys.
+
+The package variant controller will add or set a condition of type
+`ContextInjected` in the PackageVariant resource status. This will be set to
+`True` if and only if:
+- The user specified `spec.packageContext` (either or both fields).
+- The `kptfile.kpt.dev` ConfigMap exists.
+- The controller successfully modified it as specified.
+
+Otherwise, the condition should be set to `False`.
+
+If the controller is unable to modify the ConfigMap for some reason, this is
+considered an error and should prevent generation of the Draft. This will result
+in the condition `DownstreamEnsured` being set to `False`.
+
 #### KRM Function Pipeline
 
-#### Configuration Injection
+PackageVariant resource creators may specify a list of KRM functions to either
+execute on package creation and update, add to the package Kptfile, or both. The
+list of functions to execute as part of package creation or update is listed in
+`spec.execPipeline`, whereas the list of functions to ensure exist in the
+downstream Kptfile pipeline are listed in `spec.prefixPipeline` and
+`spec.suffixPipeline`. These use the Kptfile
```

Comment:
@mortent @natasha41575 This is probably way overkill and it's kinda hard. I am thinking we just add `execPipeline` for now. This means you couldn't mutate the contents of the upstream pipeline in package variant. I think that's probably OK. We may need a different primitive for that (something like mixins, perhaps).

Similarly, I am also thinking about how much flexibility I want to support along the lines of `fromField` when generating PVs from targets in PVSs. The Function type used in Pipeline is pretty complex; I don't want to have to allow every possible field to be derived fromField in there. I expect we may need the function inputs to vary. I guess I can think about that when I get to PVS design. Maybe we can get away without that.

My guess is we should do the very simplest thing first, and then see what use cases arise. The primary, initial use case for this is just to call `set-namespace` and maybe the occasional `set-labels` or `upsert-resource`.

Reply:
Ok, simplified this.